luarules/gadgets/unit_builder_priority.lua

```diff
@@ -87,219 +197,479 @@ local updateFrame = {}
+		SendToUnsynced("BuildSpeedsChanged", nCBS, unpack(changedBuildSpeeds))
+		changedBuildSpeeds = {}
+		nCBS = 0
+	end
+end
+
 
-    costID[unitID] = cost[unitDefID]
+local function UDN(unitID) --get unit name
+	return UnitDefs[Spring.GetUnitDefID(unitID)].name
 end
 
-function gadget:UnitFinished(unitID, unitDefID, teamID, builderID)
-    buildTargetOwners[unitID] = nil
-	costID[unitID] = nil
+
+function gadget:UnitCreated(unitID, unitDefID, teamID)
+	if unitBuildSpeed[unitDefID] then
+		canBuild[teamID][unitID] = true
+		maxBuildSpeed[unitID] = unitBuildSpeed[unitDefID]
+	end
+	if canPassive[unitDefID] then
+		spInsertUnitCmdDesc(unitID, cmdPassiveDesc)
+		local lowPriority = spGetUnitRulesParam(unitID, ruleName) or nil -- non-nil rule means passive
+		passiveCons[teamID][unitID] = lowPriority
+		SetBuilderPriority(unitID, lowPriority)
+
+		if VERBOSE then 
+			Spring.Echo(string.format("UnitID %i of def %s has been set to %s = %s",
+					unitID, UnitDefs[unitDefID].name, ruleName, tostring(passiveCons[teamID][unitID])))
+		end
+	end
+	--here we start tracking the unit, that is beeing built until it is finished. We use it to calculate the resources wanted by the constructors
+	costIDOverride[unitID +			   0] = cost[unitDefID]  --  
+	costIDOverride[unitID +	energyOffset] = cost[unitDefID + energyOffset]
+	costIDOverride[unitID + buildTimeOffset] = cost[unitDefID + buildTimeOffset]
 end
 
 function gadget:UnitGiven(unitID, unitDefID, newTeamID, oldTeamID)
-    if passiveCons[oldTeamID] and passiveCons[oldTeamID][unitID] then
-        passiveCons[newTeamID][unitID] = passiveCons[oldTeamID][unitID]
-        passiveCons[oldTeamID][unitID] = nil
-    end
-
-    if canBuild[oldTeamID] and canBuild[oldTeamID][unitID] then
-        canBuild[newTeamID][unitID] = true
-        canBuild[oldTeamID][unitID] = nil
-    end
+	gadget:UnitDestroyed(unitID, unitDefID, oldTeamID)
+	gadget:UnitCreated(unitID, unitDefID, newTeamID)
 end
 
 function gadget:UnitTaken(unitID, unitDefID, oldTeamID, newTeamID)
-    gadget:UnitGiven(unitID, unitDefID, newTeamID, oldTeamID)
+	gadget:UnitGiven(unitID, unitDefID, newTeamID, oldTeamID)
 end
 
 function gadget:UnitDestroyed(unitID, unitDefID, teamID)
-    canBuild[teamID][unitID] = nil
-
-    passiveCons[teamID][unitID] = nil
-    realBuildSpeed[unitID] = nil
-    currentBuildSpeed[unitID] = nil
-    buildTargetOwners[unitID] = nil
-
-    costID[unitID] = nil
+	-- Clear Team stuff
+	canBuild[teamID][unitID] = nil
+	roundRobinLastRoundTeamBuilders[teamID][unitID] = nil
+	passiveCons[teamID][unitID] = nil
+	
+	-- clear unit data
+	maxBuildSpeed[unitID] = nil
+	currentBuildSpeed[unitID] = nil
+
+	costIDOverride[unitID +			   0] = nil
+	costIDOverride[unitID +	energyOffset] = nil
+	costIDOverride[unitID + buildTimeOffset] = nil
 end
 
+function gadget:UnitFinished(unitID, unitDefID, unitTeam)
+	-- as its no longer being built, clear cache of its costs
+	costIDOverride[unitID +			   0] = nil
```

Comment:
Probably a typo here - too many spaces :)

Reply:
this is how I copied it from Beherith, Floris
;)